pkg/template/controller/readiness.go

```diff
@@ -0,0 +1,99 @@
+package controller
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/apis/batch"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	deploymentutil "k8s.io/kubernetes/pkg/controller/deployment/util"
+
+	deployapi "github.com/openshift/origin/pkg/deploy/apis/apps"
+)
+
+func checkDeploymentReadiness(obj runtime.Object) (bool, bool) {
```

Comment:
There is code all over e2e tests that basically emulates all of this multiple times.  There was an external project that tried to do this and got it wrong.

I had expected a per api group helper upstream (dealing with both internal and unversioned) which exposed a condition that took a deployment / statefulset / daemonset and returned "ready" or "not ready".  And maybe a variant that took a client for services and fetched an endpoint or something.  I then expected to layer a `kubectl wait` command around those basic libraries.  Eventually, I expected to have a subresource for resources that would encapsulate whatever we learned from `kubectl wait`.